gmso/tests/test_expression.py

```diff
@@ -0,0 +1,243 @@
+
+    def test_expression(self):
+        expression = _PotentialExpression(
+            expression='a*x+b',
+            independent_variables='x',
+            parameters={
+                'a': 1.0 * u.dimensionless,
+                'b': 2.0 * u.dimensionless
+            }
+        )
+
+        assert expression.expression == sympy.sympify('a*x+b')
+        assert 'a' in expression.parameters.keys()
+        assert 'b' in expression.parameters.keys()
+        assert expression.parameters['a'] == 1.0 * u.dimensionless
+        assert expression.parameters['b'] == 2.0 * u.dimensionless
+
+    def test_expression_multiple_indep_vars(self):
+        expression = _PotentialExpression(
+            expression='a^2+2*a*b+b^2+2*theta*phi',
+            independent_variables={'theta', 'phi'},
+            parameters={
+                'a': 2.0 * u.nm,
+                'b': 2.0 * u.rad
+            }
+        )
+        theta = sympy.Symbol('theta')
+        phi = sympy.Symbol('phi')
+        assert theta in expression.independent_variables
+        assert phi in expression.independent_variables
+        assert theta in expression.expression.free_symbols
+        assert phi in expression.expression.free_symbols
+
+    def test_invalid_expression(self):
+        with pytest.raises(ValueError) as e:
+            expression = _PotentialExpression(
+                expression='a*x+b',
+                independent_variables='x',
+                parameters={
+                    'sigma': 1.0 * u.nm,
+                    'phi': 1.0 * u.rad
+                }
+            )
+            assert 'ValueError: Missing necessary dependencies to ' \
+                   'evaluate potential expression. Missing symbols: {b, a}' in e
+
+    def test_invalid_indep_vars(self):
+        with pytest.raises(ValueError) as e:
+            expression = _PotentialExpression(
+                expression='a*x+b',
+                independent_variables='j',
+                parameters=None
+            )
+            assert "symbol j is not in expression's free symbols Cannot " \
+                   "use an independent variable that doesn't exist in the " \
+                   "expression's free symbols {x, a, b}" in e
+
+    def test_non_parametric_expression(self):
+        expression = _PotentialExpression(
+            expression='a^2+2*a*b+b^2',
+            independent_variables='a',
+            parameters=None
+        )
+        assert expression.is_parametric is False
+        with pytest.raises(AttributeError) as e:
+            assert expression.parameters
+            assert 'Object of type _PotentialExpression has no attribute parameters'
+
+    def test_set_indep_variables(self):
+        expression = _PotentialExpression(
+            expression='a^2+2*a*b+b^2',
+            independent_variables='a',
+            parameters=None
+        )
+        expression.independent_variables = {'b'}
+        assert sympy.Symbol('b') in expression.independent_variables
```

Comment:
Just want to be sure of the expected behavior, when we change the `expression.independent_variables` to `b`,  we assert below that `a` is no longer in `expression.independent_variables`, what is `a` now at this point of the code? I would expect it to be a `parameter`, but it would not have a value, nor was it specified.

Reply:
This is intended behaviour. The `_PotentialExpression` instance in this case in `non-parametric`, which means it will have no parameters. So, `a` will be just a free variable. So, let's look at this example

```py
expression = _PotentialExpression(
            'a^x + b^y + c^z',
            independent_variables={'x', 'y', 'z'},
            parameters={
                'a': 2.6 * u.nm,
                'b': 2.7 * u.nm,
                'c': 22.8 * u.hertz
            }
        )
```
Now if you are to change the `independent_variables` for the expression above, following error is raised:

```py
expression.independent_variables = {'a', 'b', 'c'}
ValueError: Missing necessary dependencies to evaluate potential expression. Missing symbols: {y, x, z}
``` 
However, if you were to do the same for non-parametric parameters=None, it doesn't complain, because there are no parameters to infer.

May be a classmethod should be added here as well, to convert non-parametric `PotentialExpression` to a parametric `PotentialExpression`.

Reply:
I was initially confused by this too (about the test and the behavior of `independent_variables` setter). Idk if you had the same confusion as I did @justinGilmer, but I initially thought that `expression.independent_variables = {'b'}` just "kicks" `a` out of the `independent_variables` list, but it turn out the setter just overwrites whatever `independent_variables` set earlier, and just check to make sure the new outcome is consistent if the parameters have already been specified.

Reply:
Ok this makes a lot more sense now! Thank you both! That behavior makes complete sense. We should probably make this an example in the docs somewhere.